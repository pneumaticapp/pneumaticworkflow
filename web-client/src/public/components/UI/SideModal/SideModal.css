@keyframes slideIn {
  0% {
    transform: translateX(100%);
  }
  100% {
    transform: translateX(0);
  }
}

@keyframes slideDown {
  0% {
    transform: translateX(0);
  }
  100% {
    transform: translateX(100%);
  }
}

.side-modal {
  position: fixed;
  z-index: 1050;
  top: 1rem;
  right: 0;
  bottom: 1rem;
  padding: 32px;
  overflow: hidden;
  width: 416px;
  max-width: calc(100% - 16px);
  color: white;
  background-color: #262522;
  border-radius: 32px 0 0 32px;
}

.side-modal_opened {
  animation: slideIn var(--menu-collapse-time) ease-out;
}

.side-modal_closed {
  animation: slideDown var(--menu-collapse-time) ease-out;
}

.side-modal_close {
  position: absolute;
  z-index: 1;
  top: 30px;
  right: 25px;
  cursor: pointer;
  background-color: transparent;
  border: 0;

  svg {
    fill: #62625f;
    transition: fill 0.3s;
  }

  &:hover svg {
    fill: white;
  }
}

.side-modal-parts {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.side-modal-parts_header {
  margin-right: 30px;
  min-height: 24px;
}

.side-modal-parts_body {
  min-height: 20px;
  flex-grow: 1;
}

.side-modal-parts_footer {
  min-height: 20px;
}
