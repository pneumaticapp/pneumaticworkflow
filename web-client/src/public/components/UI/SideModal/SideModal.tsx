/* eslint-disable */
/* prettier-ignore */
import * as React from 'react';
import classnames from 'classnames';
import * as ReactDOM from 'react-dom';
import OutsideClickHandler from 'react-outside-click-handler';

import { ClearIcon } from '../../icons';
import { history } from '../../../utils/history';

import styles from './SideModal.css';

export interface ISideModalProps {
  isClosing?: boolean;
  className?: string;
  onOpen?(): void;
  onClose(): void;
  onAfterClose?(): void;
}

export interface ISideModalPartProps {
  children?: React.ReactNode;
  className?: string;
}

const Header = (props: ISideModalPartProps) => (
  <div className={classnames(styles['side-modal-parts_header'], props.className)}>
    {props.children}
  </div>
);

const Body = (props: ISideModalPartProps) => (
  <div className={classnames(styles['side-modal-parts_body'], props.className)}>
    {props.children}
  </div>
);

const Footer = (props: ISideModalPartProps) => (
  <div className={classnames(styles['side-modal-parts_footer'], props.className)}>
    {props.children}
  </div>
);

export class SideModal extends React.Component<ISideModalProps> {
  public static Header = Header;
  public static Body = Body;
  public static Footer = Footer;

  private hanldeClickOutside = () => {
    const { onClose } = this.props;

    onClose();
  };

  private handleKeyDown = (event: KeyboardEvent) => {
    const ESC_CODE = 27;

    const { onClose } = this.props;

    if (event.keyCode === ESC_CODE) {
      onClose();
    }
  };

  public componentDidMount() {
    const { onOpen, onClose } = this.props;
    window.addEventListener('keydown', this.handleKeyDown);
    onOpen?.();

    history.listen(() => {
      onClose();
    });
  }

  public componentWillUnmount() {
    const { onAfterClose } = this.props;
    window.removeEventListener('keydown', this.handleKeyDown);
    onAfterClose?.();
  }

  private renderModal() {
    const {
      className,
      isClosing,
      onClose,
    } = this.props;

    return (
      <OutsideClickHandler onOutsideClick={this.hanldeClickOutside}>
        <div className={classnames(
          styles['side-modal'],
          !isClosing ? styles['side-modal_opened'] : styles['side-modal_closed'],
        )}>
          <button
            onClick={onClose}
            type="button"
            className={styles['side-modal_close']}
            aria-label="Close modal"
            data-test-id="close"
          >
            <ClearIcon />
          </button>

          <div className={classnames(styles['side-modal-parts'], className)}>
            {this.props.children}
          </div>
        </div>
      </OutsideClickHandler>
    );
  }

  public render() {
    return ReactDOM.createPortal(this.renderModal(), document.body);
  }
}
