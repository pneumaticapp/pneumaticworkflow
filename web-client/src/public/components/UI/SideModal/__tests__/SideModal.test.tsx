/* eslint-disable */
/* prettier-ignore */
import * as React from 'react';
import { shallow } from 'enzyme';

import { SideModal } from '../SideModal';

const mockProps = {
  onClose: jest.fn(),
};

describe('SideModal', () => {
  it('rendered correctly', () => {
    const wrapper = shallow(<SideModal {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('calls onClose when click close button', () => {
    const wrapper = shallow(<SideModal {...mockProps} />);
    const closeButton = wrapper.find('[data-test-id="close"]');
    closeButton.simulate('click');

    expect(mockProps.onClose).toHaveBeenCalled();
  });
});
