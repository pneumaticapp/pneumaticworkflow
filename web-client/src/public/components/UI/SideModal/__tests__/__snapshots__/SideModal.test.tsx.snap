// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideModal rendered correctly 1`] = `
<Portal
  containerInfo={<body />}
>
  <OutsideClickHandler
    disabled={false}
    display="block"
    onOutsideClick={[Function]}
    useCapture={true}
  >
    <div
      className="side-modal side-modal_opened"
    >
      <button
        aria-label="Close modal"
        className="side-modal_close"
        data-test-id="close"
        onClick={[MockFunction]}
        type="button"
      >
        <ClearIcon />
      </button>
      <div
        className="side-modal-parts"
      />
    </div>
  </OutsideClickHandler>
</Portal>
`;
